src/auth/idtokenclient.ts

```diff
@@ -0,0 +1,67 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {BodyResponseCallback} from '../transporters';
+
+import {Credentials} from './credentials';
+import {Headers, OAuth2Client, RequestMetadataResponse} from './oauth2client';
+
+export interface IdTokenOptions {
+  /**
+   * The client to make the request to fetch an ID token.
+   */
+  idTokenProvider: IdTokenProvider;
+  /**
+   * The audience to use when requesting an ID token.
+   */
+  targetAudience: string;
+}
+
+export interface IdTokenProvider {
+  fetchIdToken: (targetAudience: string) => Promise<string>;
+}
+
+export class IdTokenClient extends OAuth2Client {
```

Comment:
I think we'd potentially want to implement the `refreshToken` method in the `OAuth2Client` base class, but then call `options.idTokenProvider.refreshToken`?

I think this logic is darn close, but might not work right now (as the JWT client and the compute clent have slightly different token refresh logic than the base OAuth2Client class).

Reply:
I didn't want to conflate fetching ID tokens  (`fetchIdToken`) with fetching access tokens (`refreshToken`). on JWT the method is the same, but on Compute it's different (e.g. a different path is used).

It made more sense to me to implement a new method in order to:
1. make it distinct when an IdToken is being requested
2. Allow for the interface (`IdTokenProvider`) to be implemented only by the clients who are compatible